### DISCUSSION TITLE
Wish: Improve signal-to-noise ratio by setting white-list of words per feed (and also a blacklist)

### DISCUSSION THREAD

**Discussion by gustavonmartins** [spacecowboy/Feeder#402] [Ideas]
### Introduction
I really love this app, even though I just use it since 5 days. Changed my life! I don access the browser anymore, nor social media.

### Background:
I get a much better signal to noise ratio in Feeder: From about 100 articles, 15 are useful. In Social media its 100 times worse than that.

I have a simple idea to improve this ratio even more, based on my use-case. I hope it will be useful for more people.

### Issue at hand:

1. I subscribed to a local newspaper, but about half of its content is useless, and I can know it by the _title_. On this site, when they have nothing useful to talk about, they start the title with "28th October on history: What happened on that day 100 years ago". From 100 news I get from them, 50 are those type of undesired content. This could be easily eliminated by blocking the words "on history"

2. I have another source, and while most of it is noise, if some specific keywords (like "food" or "diet") are present on the _body_ of the message (not the _title_, as they dont use titles), I know I will have a great read! My idea would be that for this specific source all content not using these specific words should be blocked.

### Proposed solution:

1. Allow a _per-feed_ filter which can either block any content with a desired list of words ("on History" on my case, but could be more)

2. Allow a _per-feed_ MUST-HAVE list of words, which will be useful on my issue number 2. This would be implemented under the "Edit feed" or "Add feed" option
2.1 Observe that on this case, the keyword is inside the content o the article, not the title or subtile (this source doesnt use titles), so the filter has to be able of going into the body of the message. It can even be that the filter has to warn me that it will only work if I also select "Fetch full articles by default" for this source.

### Current capabilities close to what I need, but not quite there:

1. There is a Universal Block list under Settings which I already use to cover case 1). I believe it is ok for now and wont hide too many content from other feeds.

2. There is no list of MUST-HAVE words, not globally and not per feed, and creating one globally would be too strict. To implement one per-feed would be the desired way.

**Comment by gustavonmartins** [comment]
### Update:
I came to understand that item 1) is already being treated on https://github.com/spacecowboy/Feeder/issues/372.

Thus, I just would need the **whitelist** to be enabled on a **per-feed** basis

**Comment by gustavonmartins** [comment]
@spacecowboy If I would implement this myself, what would be the outline of what I have to do, meaning:

Which files would I probably have to change?
Which functions?
Which classes?
Is there an overview diagram of how they connect?

Are there automated tests for similar feature I could get inspiration from?

Thanks

**Reply by spacecowboy** [reply]
the existing block list feature is the only inspiration i think.

there's more than one way of doing it.

one hard requirement is that it uses the same mechanism as the existing block list feature.

I would consider adding a column in the block list sql table, referencing the feeds table and go from there.

then all the sql queries need to be updated, which can be the tricky part

**Comment by xan2622** [comment]
I second OP's second suggestions. 
It would be very convenient if the user could filter each RSS feed by adding whitelisted words or blocked words.

May I suggest to take inspiration on **MonitoRSS** (https://monitorss.xyz/ - https://github.com/synzen/monitorss) filtering system? 
It allows to finetune the filtering with conditions (I.E.: title does contain "food" and description does not contain "meat"), which is very efficient.

![image](https://github.com/user-attachments/assets/0cf51f3d-601f-41c2-8e6e-1146f9e168b2)

**Comment by valpamp** [comment]
Hi everyone, I wanted to open this feature request but I see that someone already beat me to it. I love this app and I find it to work even better than desktop-focused feed readers I use on different platforms. In my view, the ability to filter feeds is the only feature that is currently missing.

I have no idea if this is pertinent, but if I may suggest a way of implementing this functionality which may inspire someone competent to work on this, I would definitely point to [quiteRSS](https://github.com/QuiteRSS/quiterss/tree/a672609940a49776eb097abe99fa2a5b39578872/src/newsfilters)'s newsfilter module as an ideal example.
